### ISSUE TITLE
Difficult-to-reproduce child exit issue

### ISSUE THREAD

**Issue by vuori** [vuori/weasel-pageant#7]
During heavy use such as running an Ansible playbook the daemon may very rarely exit. This has been reported by a few people, but no one has managed to produce a log and there doesn't appear to be a clear way to reproduce the issue.

Tested so far:

  * SIGCHLD handling logging: doesn't appear to be related.
  * Minor nits in child restart: no effect.

Todo:

  * Stress test to run multiple agent ops in parallel to reproduce the issue.
  * Run in daily use with suitable settings to catch possible core file.
  * Implement file logging to catch issue in normal use.

**Comment by chris-morgan** [comment]
I regularly have it fail—typically several times per week, commonly multiple times in a day. I just got to the point of writing a simple alias to restart it so I can get on with my work; then it’s not a *big* bother. I’ve contemplated wrapping `ssh` with something that checks if the agent has died and restarts it, before exec'ing /usr/bin/ssh, but I haven’t done it.

If you have suggestions of things I can casually run in the background, I can help; but I’m keeping busy enough at present and have it working tolerably well enough that I’m not going to take initiative on it.

**Comment by vuori** [comment]
Ok, that's quite a lot. I have seen the issue maybe once a month at most and have never been able to catch it with debugging tools. Could you try the following:

  1. Open a new terminal window and run `strace -o trace.txt weasel-pageant -d`
  2. Open a second terminal and copy the `SSH_AUTH_SOCK` line from the first terminal to the second.
  3. Leave the first terminal running in the background. Use ssh or whatever normally in the second terminal until `weasel-pageant` dies in the first terminal.
  4. Send anything that was printed by `weasel-pageant` and the contents of `trace.txt` to me by mail. The trace file may contain names of your ssh keys and such, but won't contain private keys. Feel free to sanitize the contents as needed.

`strace` can be installed with `apt install strace` on Ubuntu.

**Comment by vuori** [comment]
It would be great if you can get a trace. Attaching strace to a normally running process (`strace -p <pid>`) and getting the output from that might be sufficient too.

**Comment by Vashiru** [comment]
Oh before I forget to add it. I'm typically using it with:

```
eval $(/c/weasel/weasel-pageant-1.1/weasel-pageant -r)
```

I think (but I've only used that for a few days) I've had less crashes using:
```
eval $(/c/weasel/weasel-pageant-1.1/weasel-pageant -r -a "/tmp/.weasel-pageant-$USER")
```

**Comment by chris-morgan** [comment]
Yeah, this is indeed amusing: I’d stopped using -a because of the recommendation with 1.1 not to use it, but went back to it as part of running it through strace, so that it was not a burden of manual work. Guess what: it hasn’t died since then.

**Comment by lusitania** [comment]
I see the same crash frequency. The strange thing is that once I have the daemon running in debug mode it stops dying on me.

This is how I call my weasel:
```
eval $(~/bin/weasel-pageant -q -H ~/win/AppData/Local/weasel-pageant/helper.exe)
```

**Comment by vuori** [comment]
It's pretty surprising that either of those switches have an effect on this. Can someone try to get a trace of a normally-invoked process (no `-a` or `-d`) exiting? Open a console normally, check `ps` for that process ID of the `weasel-pageant` running in that window and start `strace -p <process id> -o trace.txt &`, then use the terminal normally until the process exits.

**Comment by Vashiru** [comment]
@vuori I think I managed to reproduce the issue (still using -d though) a couple of days ago. It was more difficult to reproduce, but I think I managed. I opened / closed several terminal windows, switched between normal SSH sessions, jumped sessions and a few SSH tunnels to give it a run money so to speak. I've send you the trace via the e-mail listed on your github page.

**Comment by vuori** [comment]
Thanks, I'll check it out asap.

**Comment by vuori** [comment]
I have uploaded version 1.2b1 for testing. Please try that out and see if it resolves the problem.

**Comment by vuori** [comment]
Happy new year. If the beta is working for everyone I'll put out the official release this weekend.

**Comment by Vashiru** [comment]
@vuori I've installed the beta today and started using it using my day to day work (when I usually run into issues). I'll let you know how it goes.

**Comment by lusitania** [comment]
I've been working with the new version for a few days. No issues. Thanks 👏
